### PR TITLE
[AuthenticationServices] Updates for Xcode13 Beta3

### PR DESCRIPTION
**Pull request by tj-devel709**
Hi team, I am having a rough time with this binding..

Here are the Xtro errors I am currently getting:
On iOS, Mac, and MacCatalyst: https://gist.github.com/tj-devel709/b70a6205ff785d0faecede874f6f0d22
On tvOS: !unknown-type! ASAuthorizationCustomMethod bound

## The p-invoke
- !missing-pinvoke! ASAuthorizationAllSupportedPublicKeyCredentialDescriptorTransports is not bound

I attempted to create the partial class 'PublicPrivateKeyAuthentication' and added it to frameworks.sources.

- Where can I find information about what should go in the DllIimport?
- From the Xtro Error, it looks like it is not recognizing this class. Is there something else I am missing?

## The Protocol Conformance (Ignore this one for right now, need to talk to Alex about specifics in 1:1!)

- !missing-protocol-conformance! ASAuthorizationPlatformPublicKeyCredentialAssertionRequest should conform to ASAuthorizationPublicKeyCredentialAssertionRequest
- !missing-protocol-conformance! ASAuthorizationPlatformPublicKeyCredentialDescriptor should conform to ASAuthorizationPublicKeyCredentialDescriptor
- !missing-protocol-conformance! ASAuthorizationSecurityKeyPublicKeyCredentialAssertionRequest should conform to ASAuthorizationPublicKeyCredentialAssertionRequest
- !missing-protocol-conformance! ASAuthorizationSecurityKeyPublicKeyCredentialDescriptor should conform to ASAuthorizationPublicKeyCredentialDescriptor

I created empty interfaces of these protocol interfaces and tried to use them, but it doesn't seem to solve this issue.

## The Unknown Types
- !unknown-type! ASAuthorizationPublicKeyCredentialAttestationKind bound
- !unknown-type! ASAuthorizationPublicKeyCredentialResidentKeyPreference bound
- !unknown-type! ASAuthorizationPublicKeyCredentialUserVerificationPreference bound
- !unknown-type! ASAuthorizationSecurityKeyPublicKeyCredentialDescriptorTransport bound
- !unknown-type! ASAuthorizationCustomMethod

I created enums for these elements following the names xcode gives. For example, 
```
API_AVAILABLE(macos(12.0), ios(15.0)) API_UNAVAILABLE(watchos)
typedef NSString *ASAuthorizationPublicKeyCredentialAttestationKind NS_TYPED_EXTENSIBLE_ENUM;
```
but xtro doesn't seem to like these. 

Any advice on how to move forward on these would be appreciated!

### CODE REVIEW

File: src/frameworks.sources [hunk 1]
```diff
@@ -225,6 +225,7 @@ AUDIOUNIT_SOURCES = \
 AUTHENTICATIONSERVICES_SOURCES = \
 	AuthenticationServices/ASAuthorization.cs \
 	AuthenticationServices/ASAuthorizationRequest.cs \
+	PublicPrivateKeyAuthentication.cs \
```
Comment (rachelkang) [inline src/frameworks.sources:228]: I think fixing this typo should help you get closer!
```suggestion
	AuthenticationServices/PublicPrivateKeyAuthentication.cs \
```

Comment (tj-devel709) [reply src/frameworks.sources:228]: Completely missed this haha thank you!

File: src/AuthenticationServices/PublicPrivateKeyAuthentication.cs [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+namespace AuthenticationServices {
+
+    // authenticationservices.cs contain the following attributes for this class
+    // [NoWatch, NoTV, Mac (12,0), iOS (15,0), MacCatalyst (15,0)]
+	public static partial class PublicPrivateKeyAuthentication {
+        [DllImport ("__Internal")]
+        static extern /* NSString[] */ IntPtr AllSupportedPublicKeyCredentialDescriptorTransports ();
+
+        public static NSString[] GetAllSupportedPublicKeyCredentialDescriptorTransports () {
+            return NSArray.ArrayFromHandle<NSString> (AllSupportedPublicKeyCredentialDescriptorTransports ());
```
Comment (mandel-macaque) [inline src/AuthenticationServices/PublicPrivateKeyAuthentication.cs:29]: In the signature you say you return a NSString[] yet you are returning an NSArray<NSString>. Did you try to compile this?

File: src/authenticationservices.cs [hunk 3]
```diff
@@ -930,4 +952,321 @@ interface ASAccountAuthenticationModificationViewController {
...
+	[Protocol]
+	interface ASAuthorizationPublicKeyCredentialAssertion : ASPublicKeyCredential
+	{
+		[Abstract]
+		[Export ("rawAuthenticatorData", ArgumentSemantic.Copy)]
+		NSData RawAuthenticatorData { get; }
+
+		[Abstract]
+		[Export ("userID", ArgumentSemantic.Copy)]
+		NSData UserID { get; }
```
Comment (mandel-macaque) [inline src/authenticationservices.cs:1009]: I believe ID should be Id.

File: src/authenticationservices.cs [hunk 4]
```diff
@@ -930,4 +952,321 @@ interface ASAccountAuthenticationModificationViewController {
...
+
+	interface IASAuthorizationPublicKeyCredentialDescriptor { }
+
+	[NoWatch, Mac (12,0), iOS (15,0), MacCatalyst (15,0), NoTV]
+	[Protocol]
+	interface ASAuthorizationPublicKeyCredentialDescriptor : NSSecureCoding, NSCopying
+	{
+		[Abstract]
+		[Export ("credentialID", ArgumentSemantic.Copy)]
+		NSData CredentialID { get; set; }
```
Comment (mandel-macaque) [inline src/authenticationservices.cs:1047]: Same ID -> Id

File: src/authenticationservices.cs [hunk 5]
```diff
@@ -930,4 +952,321 @@ interface ASAccountAuthenticationModificationViewController {
...
+	[Protocol]
+	interface ASAuthorizationPublicKeyCredentialRegistrationRequest : NSSecureCoding, NSCopying
+	{
+		[Abstract]
+		[Export ("relyingPartyIdentifier")]
+		string RelyingPartyIdentifier { get; }
+
+		[Abstract]
+		[Export ("userID", ArgumentSemantic.Copy)]
+		NSData UserID { get; set; }
```
Comment (mandel-macaque) [inline src/authenticationservices.cs:1069]: Same.

File: src/AuthenticationServices/PublicPrivateKeyAuthentication.cs [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+using ObjCRuntime;
+
+#nullable enable
+
+namespace AuthenticationServices {
+
+    // authenticationservices.cs contain the following attributes for this class
+    // [NoWatch, NoTV, Mac (12,0), iOS (15,0), MacCatalyst (15,0)]
+	public static partial class PublicPrivateKeyAuthentication {
+        [DllImport ("__Internal")]
```
Comment (mandel-macaque) [inline src/AuthenticationServices/PublicPrivateKeyAuthentication.cs:25]: There is a Constants static class that contains all the variables needed to do the DllImport. You can see how to do it doing a git grep in the src directory
```bash
git grep DllImport
```
Which returns:
```
ARKit/ARSkeleton.cs:            [DllImport (Constants.ARKitLibrary)]
AVFoundation/AVSampleBufferExtensions.cs:        [DllImport (Constants.AVFoundationLibrary)]
AVFoundation/AVTypes.cs:                [DllImport (Constants.AVFoundationLibrary)]
Accelerate/vImageTypes.cs:              [DllImport (Constants.AccelerateImageLibrary)]
Accelerate/vImageTypes.cs:              [DllImport (Constants.AccelerateImageLibrary)]
Accelerate/vImageTypes.cs:              [DllImport (Constants.AccelerateImageLibrary)]
Accelerate/vImageTypes.cs:              [DllImport (Constants.AccelerateImageLibrary)]
Accelerate/vImageTypes.cs:              [DllImport (Constants.AccelerateImageLibrary)]
Accelerate/vImageTypes.cs:              [DllImport (Constants.AccelerateImageLibrary)]
Accelerate/vImageTypes.cs:              [DllImport (Constants.AccelerateImageLibrary,EntryPoint="vImageConvolveWithBias_ARGB8888")]
Accelerate/vImageTypes.cs:              [DllImport (Constants.AccelerateImageLibrary,EntryPoint="vImageConvolveWithBias_ARGBFFFF")]
Accelerate/vImageTypes.cs:              [DllImport (Constants.AccelerateImageLibrary)]
Accelerate/vImageTypes.cs:              [DllImport (Constants.AccelerateImageLibrary)]
Accelerate/vImageTypes.cs:              [DllImport (Constants.AccelerateImageLibrary)]
Accelerate/vImageTypes.cs:              [DllImport (Constants.AccelerateImageLibrary)]
Accelerate/vImageTypes.cs:              [DllImport (Constants.AccelerateImageLibrary)]
Accelerate/vImageTypes.cs:              [DllImport (Constants.AccelerateImageLibrary)]
Accelerate/vImageTypes.cs:              [DllImport (Constants.AccelerateImageLibrary)]
Accelerate/vImageTypes.cs:              [DllImport (Constants.AccelerateImageLibrary)]
Accelerate/vImageTypes.cs:              [DllImport (Constants.AccelerateImageLibrary)]
Accelerate/vImageTypes.cs:              [DllImport (Constants.AccelerateImageLibrary)]
```
Which shows the you should be using Constants.AuthenticationServicesLibrary

**Review by mandel-macaque — CHANGES_REQUESTED** [review]

File: src/authenticationservices.cs (hunk 3 above)
Comment (tj-devel709) [reply src/authenticationservices.cs:1009]: Good call!

File: src/AuthenticationServices/PublicPrivateKeyAuthentication.cs (hunk 2 above)
Comment (tj-devel709) [reply src/AuthenticationServices/PublicPrivateKeyAuthentication.cs:29]: No, I couldn't compile that far, other blocking issues there

File: src/AuthenticationServices/PublicPrivateKeyAuthentication.cs (hunk 6 above)
Comment (tj-devel709) [reply src/AuthenticationServices/PublicPrivateKeyAuthentication.cs:25]: @rachelkang you were right! Thanks Manuel for showing how to find this!

File: src/authenticationservices.cs [hunk 7]
```diff
@@ -930,4 +954,332 @@ interface ASAccountAuthenticationModificationViewController {
...
+		string RelyingPartyIdentifier { get; set; }
+
+		[Abstract]
+		[Export ("allowedCredentials", ArgumentSemantic.Copy)]
+		IASAuthorizationPublicKeyCredentialDescriptor[] AllowedCredentials { get; set; }
+
+		[Abstract]
+		[Export ("userVerificationPreference")]
+		// Cannot bind inside a protocol
+		// [BindAs (typeof (ASAuthorizationPublicKeyCredentialUserVerificationPreference))]
```
Comment (mandel-macaque) [inline src/authenticationservices.cs:1059]: You could write a nice extension 😈

**Comment by tj-devel709** [comment]
Closing in favor of this PR: https://github.com/xamarin/xamarin-macios/pull/12203

**Pull request closed without merging**